### PR TITLE
Add full logic for the game

### PR DESCRIPTION
**Pull request by ammarkandel**
Add full logic about the game : 
                                                  - Check which player win
                                                  - Check if the player enter a valid number
                                                  - Check for available numbers for the next player
                                                  - Display the player input into the board

### CODE REVIEW

**Review by BrianSammit — CHANGES_REQUESTED** [review]
CHANGE REQUIRED :recycle:  
Great work team!! :+1: the game is working well. :smile: 
However, there are a few changes that you have to work on. 
------------------------------------

- [x] The game logic should be in the lib directory 

Screenshot from 2021-01-14 19-05-35

- [x] Use at least two classes to encapsulate different parts of game logic.

Screenshot from 2021-01-14 19-08-43

---------------------

Made these changes and request another review. :clap: 
HAPPY CODING!! 

MB5W4D

**Review by BrianSammit — CHANGES_REQUESTED** [review]
CHANGE REQUIRED :recycle:  
I can't start the game and the lib directory is empty check your commit, it could be a problem pushing the project.

Made these changes and request another review. :clap: 
HAPPY CODING!! 

programming

**Review by LamiaSristy — CHANGES_REQUESTED** [review]
# Status: Change Required. 🔴 

Hello Team, Great work so far, 🙌 .

Your game's win logic and checking draw is working perfectly, but there is a major fault in Please check my points here: 👇

-  [x]  After any invalid input your game is switching the sign, but it should not. Like, if I am pressing 1 it is printing X, and again if I press 1, It's printing "Number 1 is an invalid number.", but if I press 2 next time it is again printing X in position 2, which should be O in position 2.
- [x] With every time the board is printing it is also printing nil, please remove it.
- [x]  All game-logic related resources (classes and/or modules) are kept in the `lib` directory. ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/OOP/milestone1/lib_dir_exist.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59565018/))
Currently, your main is responsible for many logics please remove them inside 'lib' and you can make different classes for individual work, as board, winning method, game, etc.

- [Optional] 

- [ ] Currently, your game is not taking any players name, so it is difficult for the user to play without the name, consider adding logic for taking player names, and printing it with every move.

Once you make the changes, please ask again for a re-review.
Feel free to reach me at @lamiasristy 🙋

**Review by GzimAsani — CHANGES_REQUESTED** [review]
### CHANGES REQUIRED  :stop_sign: 
### HELLO  **TEAM**  I like your project and you did a great job implementing the changes from the last review :+1:  very nice work :+1:  but unfortunately there are some requirements that you did not implement from the last reviews:
- [x]  _With every time the board is printing it is also printing nil, please remove it._
- [x] _All game-logic related resources (classes and/or modules) are kept in the lib directory. (screenshot from Pathwright) Currently, your main is responsible for many logics and includes one class, please remove it_

### Also:
- [x] Please use attr_readers/attr_writers instead of attr_accessors because by using attr_accessors you allow writing that property from outside the class. That variable can be set to anything and this means that your variables can be over-written from anywhere in the application.

- [OPTIONAL]
- [ ] _Currently, your game is not taking any players name, so it is difficult for the user to play without the name, consider adding logic for taking player names, and printing it with every move._

### **Please make the required changes and submit another review request when you are ready.**

<hr>
### if you have any question kindly reach me on slack @GzimAsani

### HAPPY CODING :computer:

**Review by kevcap — CHANGES_REQUESTED** [review]
# [STATUS] CHANGES REQUESTED ♻

### Hey, nice project so far. We only have to improve a few things before moving on:

You have `p` outside the main file, on: `check_game_result.rb`, `start_game_vars.rb`, and `start_game`;
- [x] No `puts/print` and `gets.chomp` outside `main.rb` file. ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/OOP/milestone3/only_main_print.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59565018/))
- [x] You still have logic in your main file, please note that the main should represent the view of your project, so keep in mind that you should only use it only for i/o and loops. Keep the `invalid` method outside the main file.
- [x] Here we have three errors, please fix them:
image
1- After we chose 1, you print "Type number from 1 to 9", but 1 was already taken;
2- If we chose 1 again, your project prints: "Number 1 is invalid number.", which is not true, is a valid number, what is invalid is the move. The same happens when we type a letter;
3- We have typo errors: "Number 1 is invalid number." should be: "Number 1 is **_an_** invalid ~number~ **move**." And "Type **_a_** number from 1 to 9";

⚠ Its wired that you aren't using the user interface in this project, It should ask our names and print the board with numbers to make easy to visualize the spaces to select. But since it's not explicit mandatory, I'll not ask for it;



## As always, I wish you all the very best and see you both at the top🚀


👨🏻‍💻 **TSE: Kevin Alves**

  

📫 Have any question? Better call Kevin:

<a  href="mailto:kevinalves.js@gmail.com"> gmail</a> [Twitter Follow](https://twitter.com/intent/follow?screen_name=kevinnnmn) [](https://microverse-students.slack.com/team/UT86D3AP4)

File: lib/start_game.rb [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+
+  def start_game
+    game_numbers
+    until winer
+      win_methods
+      break if winer
+
+      if @avaliable.empty?
+        p 'draw'
+        break
```
Comment (ghost) [inline lib/start_game.rb]: You should not use any function that prints to the console outside of `main.rb`. That includes `p`.

File: lib/start_game_vars.rb [hunk 2]
```diff
@@ -0,0 +1,12 @@
+module StartGameVar
+  def game_numbers
+    @all_numbers = (1..9).to_a
+    @already_used = [0]
+  end
+
+  def avaliable_inputs
+    @already_used << @player_number
+    @avaliable = @all_numbers - @already_used
+    p "These numbers are avaliable for next player #{@avaliable}"
```
Comment (ghost) [inline lib/start_game_vars.rb]: You should not use any function that prints to the console outside of `main.rb`. That includes `p`.

File: lib/check_game_result.rb [hunk 3]
```diff
@@ -0,0 +1,42 @@
+require_relative './win_vars'
+class CheckGameResult
+  def initialize(got_board)
+    @board = got_board
+  end
```
Comment (ghost) [inline lib/check_game_result.rb]: Add `attr_readers/attr_writers/attr_accessors` for the above class instance variables.

**Review by ghost — CHANGES_REQUESTED** [review]
# Status: Changes Required 🛑

## Hello team 👋,

**Great job implementing the requested changes 👏. However, there are still a few requirements needed.**

1). **As the previous TSE mentioned, when a number between 1-9  is chosen again, it should not say that that number is invalid but that the move is invalid or number is already taken.**
2). **When I play a value that is not a number, e.g d or ., I get an error: 'Number 0 is an invalid number'. Please alert the user on the right invalid move.'**
3). **I played a first good move and followed that with four bad moves. The fifth move was a good move but it didn't change the player symbol. It used the X symbol instead of O symbol. Check the screenshot below.**

 screenshot1 | screenshot2
 --------------| -------------
screenshot1 | screenshot2

4). **Game related logic like classes/modules should be added in the `lib` directory and not `main .rb` file. Please fix that.**

5). **Check more in the comments below.**

### Happy coding 🥂.

**PS: You can reach me on slack (Rose Wanjohi).**

**Review by Temmarie — APPROVED** [review]
## Status: Approved :heavy_check_mark:

### Hi @ammarkandel  ! :wave:

#### Thank you for making all of the required changes, your project is approved :clap: :clap: :clap: :tada:

**Optional**

- [ ]  When I play a value that is not a number, e.g "d, p,.-", I should get an error: "Invalid input" 

#### You may merge, good luck with your next projects! :rocket:
#### Happy Coding! :+1:
__Grace Tamara Ekunola __
[GitHub](https://github.com/temmarie) | [LinkedIn](https://www.linkedin.com/in/ekunola-grace/) | [Twitter](https://twitter.com/TemmarieW)

**Pull request merged**